### DISCUSSION TITLE
Associating start/end/step parameters with query results for proper chart drawing

### DISCUSSION THREAD

**Discussion by juliusv** [perses/perses#91] [Design Discussions]
A Prometheus query response may return samples only for a partial range of the requested graph range (e.g. because time series don't exist all the time). So to draw a graph properly with the *intended* range rather than orienting ourselves purely around the returned samples, the chart implementation needs access to the original `start` and `end` time parameters associated with each query and its results. We also need the concrete `step` parameter value to detect and fill in gaps in the returned data properly.

Currently from a chart panel when using `useRangeChartQuery()`, a Prometheus range query's time range is transformed from a relative `3h`-style range to an absolute `start` / `end` pair through this call chain:

```
useRangeChartQuery()
  usePrometheusRangeChartQuery()
    useDashboardPrometheusTimeRange()
      useDashboardTimeRange()
      toPrometheusTimeRange()   <-- memoized, gets the dashboard's time range
        toAbsoluteTimeRange()   <-- gets the current time via `new Date()` and uses it to make the time range absolute
```

This fixes the absolute time range for that hook (and a similar thing is happening for the `step` parameter). Now if in a chart panel I were to also use a similar chain of hooks to get the dashboard's current time range (and step) to be able to draw the chart with its intended time range, it would end up with a slightly different absolute time since it's going to run another `new Date()` call to absolutize the time range.

I think the most robust way for drawing a chart with its intended time range would be to store the query time parameters with the rest of the query's response data (currently stored in the DataFrames). What do people think?

**Comment by Nexucis** [comment]
If we imagine we have the button that can change the start / end time, then we could have this flow:

Change the start / end (with duration or with start / end date) -> store/transform it immediately as an start/end absolute date -> propagate these changes on every panel -> Panels only deal with absolute time range.

Does it sound acceptable for you ? Or Am I missing something ?

**Comment by juliusv** [comment]
Yeah, I generally agree! It would be nice to only compute the time parameters globally once and then pass them down to all panels, rather than having each query hook calculate / adjust them individually. That way the panel would also already have all the necessary information associated with the query that it triggers and we don't need to add that information to the query result as I suggested earlier.

There's one issue currently with this, which is that the time range is currently only absolutized in a datasource-specific way, e.g. https://github.com/perses/perses/blob/bcfe3ad507ce19942d65d45e68d6f2a549f91d31/ui/prometheus-plugin/src/model/time.ts#L32-L41 for the Prometheus datasource plugin, and on the global level of the dashboard there could be multiple datasources and not just Prometheus. So maybe we could move the absolutization of the range into the global level first in a datasource-independent way, and then individual datasources can still make adjustments to those values, but now in a deterministic / repeatable way (not relying on the current value of `new Date()`).

**Reply by Nexucis** [reply]
yep totally. I don't see the time range as something specific to the datasource itself. For me the datasource would constraint the type of the query but that's it.
So if at global level we only provide absolute time, then panel should be fine, whatever the datasource is used.

**Comment by Nexucis** [comment]
Just adding a diagram that represents the different flows that can trigger a query in the variables or in the panels.

```bash
┌────────────────────┐                          ┌───────────────────────┐
│                    │                          │                       │
│                    │                          │                       │
│      User          │                          │        Variables      │
│                    │                          │                       │
│ changes time range ├─────────────────────────►│    are recomputed     ├───────────────┐
│                    │                          │                       │               │
│                    │                          │    (query involved)   │               │
│                    │                          │                       │               │            ┌────────────────────────────┐
│                    │                          │  use the global timerange             │            │                            │
└────────────────────┘                          └──────────▲─────────────               │            │                            │
                                                           │                            │            │       Panels are refreshed │
                                                           │                            │            │                            │
                                                           │                            │            │                            │
                                                           │                            └───────────►│     It used theglobal     │
                                                           │                                         │                            │
                                                           │                                         │     time range set by the  │
                                                           │                                         │                            │
                                               ┌───────────┴────────────┐                            │     user or by the dashboard
                                               │                        │                            │                            │
                                               │                        │                            │                            │
                                               │          User          │                            └────────────────────────────┘
                                               │                        │
                                               │   changes the          │
                                               │                        │
                                               │   variable values      │
                                               │                        │
                                               │                        │
                                               │                        │
                                               └────────────────────────┘
```

**Comment by LukeTillman** [comment]
I like the idea of panels/plugins only having access to absolute time ranges (definitely makes things simpler), even if the UI shows it as a relative range. Making the syncing from a relative range in the UI -> an absolute range (to trigger refetching of things for example), feels like a much better/more explicit model, so 👍  from me.

**Comment by Nexucis** [comment]
I didn't find the way to mark this discussion as solved, since we are all agreed on what we should have regarding the start / end date.

I created the issue https://github.com/perses/perses/issues/116 to summarize this discussion.

**Reply by Nexucis** [reply]
I'm locking this discussion as it is 'solved'
